Support for Map-syntax version of project(...)

jamesonwilliams opened the pull request:
Adds support for sorting the [Map-syntax version of `project(...)`](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/dsl/DependencyHandler.html#project-java.util.Map-), e.g.,
```groovy
  implementation project(path: ':kool', configuration: 'config')
```

ZacSweers reviewed (commented):
I think this is too limited as-is and should be generalized to Gradle's map syntax if you want to add it.

Even just `path:` alone doesn't actually cover the syntax of that notation, as `implementation project(path: '...', configuration: '...')` is also legal.

If a general map syntax is supported, then you would cover both of those plus `implementation group: group, name: name, version: version, classifier: classifier, ext: extension` notation.

jamesonwilliams commented:
> I think this is too limited as-is and should be generalized to Gradle's map syntax if you want to add it.
> 
> Even just `path:` alone doesn't actually cover the syntax of that notation, as `implementation project(path: '...', configuration: '...')` is also legal.

Updated the code to handle this case in https://github.com/square/gradle-dependencies-sorter/pull/40/commits/37a938ef85010707569a3dbc87284f9f35fce891. I switched to properly detecting the map key/values in the Lexer, as opposed to the ugly Regex hack I had.

> If a general map syntax is supported, then you would cover both of those plus `implementation group: group, name: name, version: version, classifier: classifier, ext: extension` notation.

The key/value pairs to `project(...)` and `implementation` are semantic w.r.t. to sort. For `project(...)`, `path` is all that matters (or at least, should be the primary sort key.) Likewise, for `implementation`, there is a specific precedence to the keys when sorting. So, I don't think we can ever rely on a generalized map syntax - rather, we'll have to encode specific rules by key based on the Gradle DSL. Thus, it could be best to tackle `implementation`'s Map notation as a follow-up.

ZacSweers commented on sort/src/main/kotlin/com/squareup/sort/DependencyComparator.kt:
```diff
@@ -72,4 +66,22 @@ internal class DependencyComparator(
+      isFileDependency() -> dependency.fileDependency().ID().text
+      else -> dependency.externalDependency().ID().text
+    }
+    // Colons should sort "higher" than hyphens. The comma's ascii value is 44, the hyphen's is
+    // 45, and the colon's is 58. We replace colons with commas and then rely on natural sort order
+    // from there. Similarly, single and double quotes have different values, but we don't care
+    // about that for our purposes.
```
Could you add a small example of this in this comment?

jamesonwilliams replied:
Sure, can do -

To be sure, though, this is just existing code/comments that I've relocated from here, without modification: https://github.com/square/gradle-dependencies-sorter/blob/b4d6f06bfde50680ad694769312de583e8fe0444/sort/src/main/kotlin/com/squareup/sort/DependencyComparator.kt#L44-L48

The original work was done in this first commit in the repo: https://github.com/square/gradle-dependencies-sorter/commit/ce5d0c93016e559a5c606760eb0ad7f47af8cb11#diff-962ccc2459401e773b32a9f3c472f40a3cc9a9c3f5b2d652a7ac955c793a3881R44-R47, with a minor and more-recent update from @autonomousapps here: https://github.com/square/gradle-dependencies-sorter/commit/c7c7d12d0cc6f1d96fcea4235b6b9de5c1911dac#diff-962ccc2459401e773b32a9f3c472f40a3cc9a9c3f5b2d652a7ac955c793a3881R46-L48

jamesonwilliams replied:
Sure - done in https://github.com/square/gradle-dependencies-sorter/pull/40/commits/9ef59c5ddac0657f7d4f98e3bf8beafd03658ba8.

This was an existing comment that I just relocated when I created this extension function, but hopefully, this leaves it in a more clear state. Let me know what you think? cc: @autonomousapps
